subprojects/core/src/test/groovy/org/gradle/internal/classpath/CallInterceptingMetaClassTest.groovy

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath
+
+import org.gradle.internal.classpath.intercept.CallInterceptorResolver
+import org.gradle.internal.classpath.intercept.CallInterceptorsSet
+import spock.lang.Specification
+
+import static org.gradle.internal.classpath.BasicCallInterceptionTestInterceptorsDeclaration.GROOVY_GENERATED_CLASS
+import static org.gradle.internal.classpath.InstrumentedGroovyCallsTracker.CallKind.GET_PROPERTY
+import static org.gradle.internal.classpath.InstrumentedGroovyCallsTracker.CallKind.INVOKE_METHOD
+import static org.gradle.internal.classpath.InstrumentedGroovyCallsTracker.CallKind.SET_PROPERTY
+
+class CallInterceptingMetaClassTest extends Specification {
+
+    CallInterceptorResolver callInterceptors = new CallInterceptorsSet(GroovyCallInterceptorsProvisionTools.getInterceptorsFromClass(GROOVY_GENERATED_CLASS).stream())
+    private MetaClass originalMetaClass = null
+    private static InterceptorTestReceiver instance = null
+
+    // We don't want to interfere with other tests that modify the meta class
+    def interceptorTestReceiverClassLock = new ClassBasedLock(InterceptorTestReceiver)
```

Comment:
❓ Would it make sense to move this field plus the `setup()` and `cleanup()` logic to a trait and reuse in all tests?